### PR TITLE
Update security best practice - brute-force protection

### PR DESCRIPTION
**Pull request by animir**
1. Move a point about brute-force protection upper as it is well known attack vector. Community should pay attention to that with examples how to achieve better security on authorisation.
2. Remove `strongloop/microgateway` as it is not well supported/used package nowdays.

### CODE REVIEW

File: en/advanced/best-practice-security.md [hunk 1]
```diff
@@ -152,6 +153,16 @@ app.use(session({
 }))
 ```
 
+## Prevent brute-force attacks against authorisation
```
Comment (crandmck) [inline en/advanced/best-practice-security.md]: "authorization"
Spelling it with an "s" is the British spelling, and by convention we use American spelling

File: en/advanced/best-practice-security.md [hunk 2]
```diff
@@ -152,6 +153,16 @@ app.use(session({
...
+
+Make sure login endpoints are protected to make private data more secure.
+
+One simple and in the same time quite powerful way is to block authorisation attempts by 2 metrics:
```
Comment (crandmck) [inline en/advanced/best-practice-security.md]: A simple and powerful technique is to block authorization attempts using two metrics:

File: en/advanced/best-practice-security.md [hunk 3]
```diff
@@ -152,6 +153,16 @@ app.use(session({
...
+1. The first is number of consecutive failed attempts by Username and IP pair.
```
Comment (crandmck) [inline en/advanced/best-practice-security.md]: ...consecutive failed attempts by the same user name and IP address.  

Note spelling of "username".

File: en/advanced/best-practice-security.md [hunk 4]
```diff
@@ -152,6 +153,16 @@ app.use(session({
 }))
 ```
 
+## Prevent brute-force attacks against authorisation
+
+Make sure login endpoints are protected to make private data more secure.
+
+One simple and in the same time quite powerful way is to block authorisation attempts by 2 metrics:
+1. The first is number of consecutive failed attempts by Username and IP pair. 
+1. The second is number of failed attempts from IP per long period of time. For example, block IP on 100 failed attempts per day.
```
Comment (crandmck) [inline en/advanced/best-practice-security.md]: ...attempts from an IP address over some long period of time. For example, block an IP address if it makes 100 failed attempts in one day.

File: en/advanced/best-practice-security.md [hunk 5]
```diff
@@ -152,6 +153,16 @@ app.use(session({
...
+
+[rate-limiter-flexible](https://github.com/animir/node-rate-limiter-flexible) package provides tools to make that easy and fast. You can find [an example of brute-force protection in docs](https://github.com/animir/node-rate-limiter-flexible/wiki/Overall-example#login-endpoint-protection)
```
Comment (crandmck) [inline en/advanced/best-practice-security.md]: The [rate-limiter-flexible](https://github.com/animir/node-rate-limiter-flexible) package provides tools to make this technique easy and fast. You can find [an example of brute-force protection in the documentation](https://github.com/animir/node-rate-limiter-flexible/wiki/Overall-example#login-endpoint-protection).

**Review by crandmck — CHANGES_REQUESTED** [review]

**Comment by crandmck** [comment]
Thanks for the PR.  I made a number of comments, mostly wording, spelling, grammar, and the like.  Please make these changes, and then we can land this, if there are no objections.  

Also, I don't have a real problem removing mention of  StrongLoop microgateway, but actually it does appear to have lots of recent commits: https://github.com/strongloop/microgateway/commits/master so perhaps we should keep it?  

I can't comment on the technical merits of`node-rate-limiter-flexible`, but the intent seems reasonable.  @dougwilson @wesleytodd Any opinions?

**Comment by animir** [comment]
@crandmck Thank you, I've made changes.

I'll share my thoughts regarding microgateway, I hope, it helps to make decision:
Microgateway got several commits for the last 2 months. As I see, somebody just added `app_name` to analytics. They were silent for the 6 months before. Many PRs made by automatic Snyk, nobody seems care. Microgateway is downloaded 1k times per week, it is not good sign too.
Btw, Strongloop puts much more efforts to https://github.com/strongloop/loopback-next

If somebody opens this security best practice page on expressjs.com, it is expected to provide useful advice. If it is developer who wants to get some knowledge, and our advice `you should learn some other complex thing to make it right`, it isn't good advice to use Microgateway just because it provides policy for rate-limiting.

**Comment by crandmck** [comment]
Thanks @animir.  Yeah, I'm fine with removing micro-gateway. 

So, AFAIC this is good to go, but I would like to get a nod from Doug or Wes before landing it.

**Review by crandmck — APPROVED** [review]

**Comment by dougwilson** [comment]
LGTM though `req.connection.remoteAddress` should probably be `req.ip`

**Comment by animir** [comment]
@dougwilson thanks for the comment. I'd like to keep example focused on certain topic.
If developer wants to use `req.ip`, he should set `trust proxy` and understand what is behind that.

**Comment by dougwilson** [comment]
The example is how to use with express. You should only ever get the ip from req.ip in express. If you're not willing to take simple feedback about this, i don't think we should link to this module from the official expressjs.com web site, imo.

**Comment by dougwilson** [comment]
This is the example i am referring to, the express middleware specific thing https://github.com/animir/node-rate-limiter-flexible/wiki/Express-Middleware

**Comment by animir** [comment]
@dougwilson Ah, here was some misunderstanding. I thought about proxy settings.
I see now, that you mean that `req.ip` is better in context of expressjs, as it is expressjs API. 

I've changed it everywhere in expressjs related examples.

**Comment by crandmck** [comment]
OK, thx for your input @dougwilson 
And thx for the PR @animir

**Pull request merged**